### ISSUE TITLE
Native Pseudo-Huber loss support

### ISSUE THREAD

**Issue by LionOrCatThatIsTheQuestion** [dmlc/xgboost#5479]
Would it be possible to support Pseudo-Huber-loss (https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function) natively?

I implemented it as a custom losss function (I use the Python SKLearn API)
``` python
def huber_approx_obj(y_true, y_pred):
    z = y_pred - y_true
    delta = 1
    scale = 1 + (z/delta)**2
    scale_sqrt = np.sqrt(scale)
    grad = z/scale_sqrt
    hess = 1/(scale*scale_sqrt)
    return grad, hess
```
but the feature importance plots don't support custom loss functions (and it slows the learning process in comparison to 'reg:squarederror'). 

The basic problem is the need for a robust regession objective; MSE can be sensitive to outliers in application.

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion Would you like to make a PR for this?  It should just be a simple class defined in `src/objective/regression_loss.h`.  I am happy to help. ;-)

**Comment by LionOrCatThatIsTheQuestion** [comment]
@trivialfis what evaluation metric should I use, rmlse okay?

Here is my code so far:

```
struct PseudoHuberError {

  XGBOOST_DEVICE static bst_float PredTransform(bst_float x) { 
    return x; 
  }

  XGBOOST_DEVICE static bool CheckLabel(bst_float label) {
    return true;
  }
 
  XGBOOST_DEVICE static bst_float FirstOrderGradient(bst_float predt, bst_float label) {
    const float z = predt - label;
    const float scale_sqrt = std::sqrt(1 + std::pow(z,2));
    return z/scale_sqrt;
  }
 
  XGBOOST_DEVICE static bst_float SecondOrderGradient(bst_float predt, bst_float label) {
    const float scale = 1 + std::pow(predt - label,2);
    const float scale_sqrt = std::sqrt(scale);
    return 1/(scale*scale_sqrt);
  }
 
  static bst_float ProbToMargin(bst_float base_score) { 
    return base_score; 
  }
 
  static const char* LabelErrorMsg() {
    return "";
  }
 
  static const char* DefaultEvalMetric() { 
    return "rmsle"; 
  }
 
  static const char* Name() { 
    return "reg:pseudohubererror"; 
  }
};
```

**Comment by trivialfis** [comment]
Does it make sense to use Pseudo-Huber loss as a metric?

**Comment by LionOrCatThatIsTheQuestion** [comment]
Guess Pseudo-Huber loss would be an option too (seems natrual to choose the same mectric as loss function?) or MAE.

The idea was to implemented Pseudo-Huber loss as a twice differentiable approximation of MAE, so on second thought MSE as metric kind of defies the original purpose.

**Comment by LionOrCatThatIsTheQuestion** [comment]
The advantage of MAE (and also MSE), is that they are better/natural interpretable. Pseudo-Huber loss does not have the same values as MAE in the case "abs(y_pred - y_true) > 1", it just has the same linear shape as opposed to quadratic.

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion We can set the default metric to be huber, as users can specify other metrics if they like.  To me using huber as the default metric seems appropriate here.  You can add a metric in `src/metric/elementise_metric.cu`.  Feel free to ping me if you have any issue around the code base.

**Comment by LionOrCatThatIsTheQuestion** [comment]
@trivialfis Could you explain what the function GetFinal(...) does? I used MAE as reference:

`struct EvalRowPHE {
  char const *Name() const {
    return "phe";
  }
  XGBOOST_DEVICE bst_float EvalRow(bst_float label, bst_float pred) const {
    bst_float diff = label - pred;
    return std::sqrt(1 + diff * diff) - 1;
  }
  static bst_float GetFinal(bst_float esum, bst_float wsum) {
    return wsum == 0 ? esum : esum / wsum;
  }
}; `

**Comment by trivialfis** [comment]
@LionOrCatThatIsTheQuestion Is there any reason we should fix `\delta` to be 1 ?

**Comment by trivialfis** [comment]
For normal cases `GetFinal` is just a weighted mean.  1 special case is the gamma deviance, which is weighted deviance: https://rdrr.io/cran/MetricsWeighted/man/deviance_gamma.html

**Comment by LionOrCatThatIsTheQuestion** [comment]
\delta should be 1 by default, but adjustable would be better than fixed - the question is more if its possible and how to implement an additional parameter for a metric?

e.g. in sklearn interface, I would just use the keyword 'reg:pseudohubererror' to specify the metric

**Comment by QuantHao** [comment]
> Would it be possible to support Pseudo-Huber-loss (https://en.wikipedia.org/wiki/Huber_loss#Pseudo-Huber_loss_function) natively?
> 
> I implemented it as a custom losss function (I use the Python SKLearn API)
> 
> ```python
> def huber_approx_obj(y_true, y_pred):
>     z = y_pred - y_true
>     delta = 1
>     scale = 1 + (z/delta)**2
>     scale_sqrt = np.sqrt(scale)
>     grad = z/scale_sqrt
>     hess = 1/(scale*scale_sqrt)
>     return grad, hess
> ```
> 
> but the feature importance plots don't support custom loss functions (and it slows the learning process in comparison to 'reg:squarederror').
> 
> The basic problem is the need for a robust regession objective; MSE can be sensitive to outliers in application.

Hi, is it possible to relax the constrain of delta equals 1 so that user could choose other delta such as 1.35 to obtain achieve 95% statistical efficiency? Or maybe just set its default as 1.35 to be compatible with sklearn?

Reference: https://scikit-learn.org/stable/modules/linear_model.html#huber-regression

**Comment by pidefrem** [comment]
> \delta should be 1 by default, but adjustable would be better than fixed - the question is more if its possible and how to implement an additional parameter for a metric?
> 
> e.g. in sklearn interface, I would just use the keyword 'reg:pseudohubererror' to specify the metric

Passing an additional parameter for a metric is done for poissonregression or tweedie regression for example. See https://github.com/dmlc/xgboost/blob/03cd087da180b7dff21bd8ef34997bf747016025/src/objective/regression_obj.cu#L174

Making `delta` a parameter would imply some refactoring because PseudoHuberError has only static member functions: https://github.com/dmlc/xgboost/blob/03cd087da180b7dff21bd8ef34997bf747016025/src/objective/regression_loss.h#L101

**Comment by trivialfis** [comment]
Reopening as a reminder.

**Comment by chan4899** [comment]
Any plans on making Delta adjustable? Turns out PseudoHuberLoss is very effective against outliers and I would like to tune delta further to get better results.
